### PR TITLE
Use node IP for advertised.listeners when node port is used

### PR DESCRIPTION
**Pull request by bartam1**
| Q               | A      |
| --------------- | ------ |
| Bug fix?        | no |
| New feature?    | yes |
| API breaks?     | no |
| Deprecations?   | no |
| Related tickets | mentioned in #901 and #900 |
| License         | Apache 2.0 |

### What's in this PR?

Using node ip based on configuration (NodePortNodeAddressType) for brokers in advertised.listeners when hostnameOverride and nodePortExternalIP are not set

### Why?

Users dont have to specify the node external addresses by hand for the brokers when nodeport is used.

### Checklist

- [x] Implementation tested
- [x] Error handling code meets the [guideline](https://github.com/banzaicloud/developer-guide/blob/master/docs/coding-style/error-handling-guide.md)
- [x] Logging code meets the guideline
- [x] User guide and development docs updated (if needed)

### CODE REVIEW

File: pkg/resources/kafka/kafka.go [hunk 1]
```diff
@@ -1161,7 +1161,15 @@ func (r *Reconciler) createExternalListenerStatuses(log logr.Logger) (map[string
...
+							if err != nil {
+								log.Error(err, fmt.Sprintf("could not get the (%s) address of the broker's (ID: %d) node for external listener (%s) configuration",
+									bConfig.NodePortNodeAddressType, broker.Id, eListener.Name))
+							}
+						} else {
+							brokerHost = bConfig.NodePortExternalIP[eListener.Name]
+						}
 					} else {
 						brokerHost = fmt.Sprintf("%s-%d-%s.%s%s", r.KafkaCluster.Name, broker.Id, eListener.Name, r.KafkaCluster.Namespace, brokerHost)
 					}
```
Comment (panyuenlau) [inline pkg/resources/kafka/kafka.go:1175]: **Suggestion**: I'd prefer to wrap all these if-else clauses with a util function, for example:
```golang
broekrHost := getBrokerHost(host)
```

The reason behind this is that the function `createExternalListenerStatuses` is already very cumbersome and difficult for others to understand all the logics in the function. Wrapping the if-else clauses would help improve the readability of this `createExternalListenerStatuses` function and make it easier for others to maintain in the future

File: pkg/resources/kafka/kafka.go [hunk 2]
```diff
@@ -1217,6 +1225,44 @@ func (r *Reconciler) getK8sAssignedNodeport(log logr.Logger, eListenerName strin
...
+	case len(podList.Items) == 0:
+		return "", fmt.Errorf("could not find pod with matching brokerId: %d in namespace '%s'", brokerId, r.KafkaCluster.Namespace)
+	case len(podList.Items) > 1:
+		return "", fmt.Errorf("multiple pod found with brokerId: %d in namespace '%s'", brokerId, r.KafkaCluster.Namespace)
+	default:
+		return r.getK8sNodeIp(podList.Items[0].Spec.NodeName, nodeAddressType)
+	}
+}
+
+func (r *Reconciler) getK8sNodeIp(nodeName string, nodeAddressType string) (string, error) {
```
Comment (panyuenlau) [inline pkg/resources/kafka/kafka.go:1247]:
```suggestion
func (r *Reconciler) getK8sNodeIP(nodeName string, nodeAddressType string) (string, error) {
```
According to this: https://github.com/golang/go/wiki/CodeReviewComments#initialisms

File: pkg/resources/kafka/kafka.go [hunk 3]
```diff
@@ -1217,6 +1225,44 @@ func (r *Reconciler) getK8sAssignedNodeport(log logr.Logger, eListenerName strin
...
+		client.MatchingLabels(apiutil.MergeLabels(apiutil.LabelsForKafka(r.KafkaCluster.Name), map[string]string{v1beta1.BrokerIdLabelKey: strconv.Itoa(int(brokerId))})),
+	); err != nil {
+		return "", err
+	}
+
+	switch {
+	case len(podList.Items) == 0:
+		return "", fmt.Errorf("could not find pod with matching brokerId: %d in namespace '%s'", brokerId, r.KafkaCluster.Namespace)
+	case len(podList.Items) > 1:
+		return "", fmt.Errorf("multiple pod found with brokerId: %d in namespace '%s'", brokerId, r.KafkaCluster.Namespace)
```
Comment (panyuenlau) [inline pkg/resources/kafka/kafka.go:1241]:
```suggestion
		return "", fmt.Errorf("multiple pods found with brokerId: %d in namespace '%s'", brokerId, r.KafkaCluster.Namespace)
```

**Review by panyuenlau — APPROVED** [review]
Implementation LGTM. 

It's just that I would really like to see those nested if-else clauses to be wrapped with a util function rather having them to live directly in the `createExternalListenerStatuses` func

**Review by Kuvesz — COMMENTED** [review]
Looks good to me, though I'd live to see @panyuenlau 's suggestions before approving.

**Review by Kuvesz — APPROVED** [review]
LGTM

**Review by panyuenlau — APPROVED** [review]
Thanks for taking it over from the community. 

Should we also update this [banzaicloud_v1beta1_kafkacluster.yaml](https://github.com/banzaicloud/koperator/blob/master/config/samples/banzaicloud_v1beta1_kafkacluster.yaml) to reflect the new changes on the KafkaCluster changes? And perhaps the [node port section in the koperator docs](https://banzaicloud.com/docs/supertubes/kafka-operator/external-listener/#nodeport) as well

File: pkg/resources/kafka/kafka.go [hunk 4]
```diff
@@ -1217,6 +1235,44 @@ func (r *Reconciler) getK8sAssignedNodeport(log logr.Logger, eListenerName strin
...
+	addressMap := make(map[string]string)
+	for _, address := range node.Status.Addresses {
+		addressMap[string(address.Type)] = address.Address
+	}
+
+	if val, ok := addressMap[nodeAddressType]; ok && val != "" {
+		return val, nil
+	}
+
+	return "", fmt.Errorf("Could not find the node (%s)'s (%s) address", nodeName, nodeAddressType)
```
Comment (panyuenlau) [inline pkg/resources/kafka/kafka.go:1273]:
```suggestion
	return "", fmt.Errorf("could not find the node (%s)'s (%s) address", nodeName, nodeAddressType)
```

See https://github.com/golang/go/wiki/CodeReviewComments#error-strings

**Review by Kuvesz — APPROVED** [review]

File: api/v1beta1/kafkacluster_types.go [hunk 5]
```diff
@@ -194,6 +194,13 @@ type BrokerConfig struct {
 	// type service to expose the broker outside the Kubernetes cluster. Also, when "hostnameOverride" field of the external listener is set
 	// it will override the broker's external listener advertise address according to the description of the "hostnameOverride" field.
 	NodePortExternalIP map[string]string `json:"nodePortExternalIP,omitempty"`
+	// When "hostNameOverride" and brokerConfig.nodePortExternalIP are empty and NodePort access method is selected for an external listener
+	// the NodePortNodeAdddressType defines the Kafka broker's Kubernetes node's address type that shall be used in the advertised.listeners property.
+	// https://kubernetes.io/docs/concepts/architecture/nodes/#addresses
+	// The NodePortNodeAddressType's values can be Hostname, ExternalIP, InternalIP, InternalDNS,ExternalDNS
+	// +kubebuilder:validation:Enum=Hostname;ExternalIP;InternalIP;InternalDNS;ExternalDNS
+	// +optional
+	NodePortNodeAddressType corev1.NodeAddressType `json:"nodePortNodeAddressType,omitempty"`
```
Comment (Kuvesz) [inline api/v1beta1/kafkacluster_types.go:203]: This should be removed from here after merging #942 right?

Comment (panyuenlau) [reply api/v1beta1/kafkacluster_types.go:203]: The API changes will be gone from this PR after #942 is merged the this PR is rebased with master

Comment (Kuvesz) [reply api/v1beta1/kafkacluster_types.go:203]: fair

**Review by Kuvesz — APPROVED** [review]

File: go.mod [hunk 6]
```diff
@@ -1,6 +1,6 @@
 module github.com/banzaicloud/koperator
 
-go 1.19
+go 1.18
```
Comment (panyuenlau) [inline go.mod:3]: Is this change intended?

Comment (bartam1) [reply go.mod:3]: No

**Review by Kuvesz — APPROVED** [review]

**Review by pregnor — APPROVED** [review]
LGTM at a glance
